service_configuration_lib/spark_config.py

```diff
@@ -74,6 +75,7 @@
 
 SUPPORTED_CLUSTER_MANAGERS = ['kubernetes', 'local']
 DEFAULT_SPARK_RUN_CONFIG = '/nail/srv/configs/spark.yaml'
+PREFERRED_SPARK_UI_PORT = 39091
```

Comment:
Reason for picking-up the specific UI port, instead of random? 
Is it because it's hard to know the random port beforehand?

Reply:
This port is specified by the pod annotation when creating a jupyter server / spark driver pod, to let Prometheus to know what the scrape endpoint is.
The reason of fixing this port instead to dynamically assign it & add pod annotation on-the-fly is because we don't need to give user permission for modifying pod info which may lead to security concerns.